internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,222 @@
+	if err != nil {
+		if errors.Is(err, contract.ErrFieldNotFound) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform preflight checks: failed to get the version of ControlPlane %s", cpKlogRef)
+	}
+	cpSemver, err := semver.ParseTolerant(*cpVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform preflight checks: failed to parse ControlPlane version %s", *cpVersion)
+	}
+
+	errList := []error{}
+
+	// Run the control-plane-stable preflight check.
+	if !skipped.Has(clusterv1.MachineSetPreflightCheckControlPlaneStable) {
+		if checkResult, err := r.controlPlaneStablePreflightCheck(ctx, controlPlane, ms); err != nil || !checkResult.IsZero() {
+			result = util.LowestNonZeroResult(result, checkResult)
+			errList = append(errList, err)
+		}
+	}
+
+	// Check the version skew policies only if version is defined in the MachineSet.
+	if ms.Spec.Template.Spec.Version != nil {
+		msVersion := *ms.Spec.Template.Spec.Version
+		msSemver, err := semver.ParseTolerant(msVersion)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to perform preflight checks: failed to parse MachineSet version %s", msVersion)
+		}
+
+		// Run the kubernetes-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubernetesVersion) {
+			if checkResult := r.kubernetesVersionPreflightCheck(ctx, cpSemver, msSemver, ms); !checkResult.IsZero() {
+				result = util.LowestNonZeroResult(result, checkResult)
+			}
+		}
+
+		// Run the kubeadm-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubeadmVersion) {
+			if checkResult, err := r.kubeadmVersionPreflightCheck(ctx, cpSemver, msSemver, ms); err != nil || !checkResult.IsZero() {
+				result = util.LowestNonZeroResult(result, checkResult)
+				errList = append(errList, err)
+			}
+		}
+	}
+
+	return result, errors.Wrap(kerrors.NewAggregate(errList), "failed to perform preflight checks")
+}
+
+func (r *Reconciler) controlPlaneStablePreflightCheck(ctx context.Context, controlPlane *unstructured.Unstructured, ms *clusterv1.MachineSet) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// Check that the control plane is not provisioning.
+	isProvisioning, err := contract.ControlPlane().IsProvisioning(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q preflight check: failed to check if ControlPlane %s is provisioning", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef)
+	}
+	if isProvisioning {
+		log.Info(fmt.Sprintf("Preflight check %q failed: ControlPlane %s is provisioning", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef))
+		conditions.MarkFalse(ms, clusterv1.MachinesCreatedCondition, clusterv1.PreflightCheckFailedReason, clusterv1.ConditionSeverityError, "Control plane is provisioning")
```

Comment:
I think we have a problem that the prechecks are overwriting each other when multiple of them fail.

WDYT about instead of logging, requeuing + marking the condition here we return some sort of precheck result up one level, and then we can aggregate the results and write the condition once

Reply:
I think this would be also good in general for the underlying PR.

This would have the nice effect that we have the aggregation + the requeu logic only in runPreflightChecks and not on the deeper layers.

WDYT?

Reply:
This was probably left behind from when we were originally returning as soon as one of the preflight checks failed. 
I will adjust this to have the aggregated result.